File: packages/benchmarks/src/runner.ts
import { exactMatch, evaluateRetrieval, tokenF1 } from "./metrics.js";
import { benchmarkFixtureSchema } from "./schema.js";
import type { BenchmarkFixture, BenchmarkReport, Retriever, RetrievedMemory } from "./types.js";

function terms(text: string): Set<string> {
  return new Set(text.toLowerCase().replace(/[^\p{L}\p{N}]+/gu, " ").split(/\s+/).filter((t) => t.length > 2));
}

export const lexicalRetriever: Retriever = (query, fixture, opts) => {
  const q = terms(query.text);
  return fixture.memories
    .filter((m) => !m.supersededBy)
    .map((m): RetrievedMemory => {
      const mt = terms(m.text);
      let overlap = 0;
      for (const token of q) if (mt.has(token)) overlap += 1;
      const exactExpected = query.expectedAnswer && m.text.toLowerCase().includes(query.expectedAnswer.toLowerCase()) ? 100 : 0;
      return { id: m.id, text: m.text, score: exactExpected + overlap / Math.max(q.size, 1), answer: query.expectedAnswer && m.text.toLowerCase().includes(query.expectedAnswer.toLowerCase()) ? query.expectedAnswer : undefined };
    })
    .sort((a, b) => (b.score ?? 0) - (a.score ?? 0) || a.id.localeCompare(b.id))
    .slice(0, opts.k);
};

function percentile(values: number[], p: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = Math.min(sorted.length - 1, Math.ceil((p / 100) * sorted.length) - 1);
  return sorted[idx] ?? 0;
}

export async function runBenchmark(
  rawFixture: BenchmarkFixture,
  retriever: Retriever,
  opts: { kValues?: number[] } = {},
): Promise<BenchmarkReport> {
  const fixture = benchmarkFixtureSchema.parse(rawFixture) as BenchmarkFixture;
  const kValues = opts.kValues ?? [1, 3, 5, 10];
  const maxK = Math.max(...kValues);
  const cases = [];
  for (const query of fixture.queries) {
    const started = performance.now();
    const retrieved = await retriever(query, fixture, { k: maxK });
    const latencyMs = performance.now() - started;
    cases.push({
      queryId: query.queryId,
      category: query.category,
      relevantIds: query.relevantMemoryIds,
      forbiddenIds: query.forbiddenMemoryIds ?? [],
      retrievedIds: retrieved.map((r) => r.id),
      answer: retrieved.find((r) => r.answer)?.answer ?? retrieved[0]?.answer,
      expectedAnswer: query.expectedAnswer,
      latencyMs,
    });
  }
  const retrieval = evaluateRetrieval(cases.map((c) => ({ queryId: c.queryId, relevantIds: c.relevantIds, retrievedIds: c.retrievedIds })), kValues);
  const answerable = cases.filter((c) => c.expectedAnswer);
  const answeredCount = answerable.length;
  const answer = {
    exactMatch: answeredCount === 0 ? 0 : answerable.reduce((sum, c) => sum + exactMatch(c.answer, c.expectedAnswer), 0) / answeredCount,
    tokenF1: answeredCount === 0 ? 0 : answerable.reduce((sum, c) => sum + tokenF1(c.answer, c.expectedAnswer), 0) / answeredCount,
    answeredCount,
  };
  const forbiddenHitRateAtK: Record<number, number> = {};
  for (const k of kValues) {
    forbiddenHitRateAtK[k] = cases.length === 0 ? 0 : cases.reduce((sum, c) => {
      const forbidden = new Set(c.forbiddenIds);
      return sum + (c.retrievedIds.slice(0, k).some((id) => forbidden.has(id)) ? 1 : 0);
    }, 0) / cases.length;
  }
  const latencies = cases.map((c) => c.latencyMs);
  return {
    fixture: { name: fixture.name, kind: fixture.kind, version: fixture.version, queryCount: fixture.queries.length, memoryCount: fixture.memories.length },
    retrieval,
    answer,
    safety: { forbiddenHitRateAtK },
    latency: {
      averageMs: latencies.length === 0 ? 0 : latencies.reduce((a, b) => a + b, 0) / latencies.length,
      p95Ms: percentile(latencies, 95),
      maxMs: Math.max(0, ...latencies),
    },
    cases,
  };
}

File: go/internal/bench/runner.go
package bench

// The fixture runner and the built-in lexical retriever, transcribed
// from packages/benchmarks/src/runner.ts.

import (
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

func itoa(n int) string { return strconv.Itoa(n) }

var termSplit = regexp.MustCompile(`[^\p{L}\p{N}]+`)

// terms is the retriever's tokenizer: lowercase, split on anything that
// is not a letter or digit, and keep tokens longer than two characters.
func terms(text string) map[string]bool {
	out := map[string]bool{}
	for _, t := range strings.Fields(termSplit.ReplaceAllString(strings.ToLower(text), " ")) {
		if len([]rune(t)) > 2 {
			out[t] = true
		}
	}
	return out
}

// LexicalRetriever is the dependency-free baseline: term overlap, with a
// large bonus when a memory literally contains the expected answer. It
// needs no model, which is what makes it usable as a CI smoke gate.
func LexicalRetriever(q Query, f Fixture, k int) ([]Retrieved, error) {
	qt := terms(q.Text)
	out := make([]Retrieved, 0, len(f.Memories))
	for _, m := range f.Memories {
		if m.SupersededBy != "" {
			continue
		}
		mt := terms(m.Text)
		overlap := 0
		for token := range qt {
			if mt[token] {
				overlap++
			}
		}
		containsAnswer := q.ExpectedAnswer != "" &&
			strings.Contains(strings.ToLower(m.Text), strings.ToLower(q.ExpectedAnswer))
		exactExpected := 0.0
		answer := ""
		if containsAnswer {
			exactExpected = 100
			answer = q.ExpectedAnswer
		}
		denom := len(qt)
		if denom < 1 {
			denom = 1
		}
		out = append(out, Retrieved{
			ID:     m.ID,
			Text:   m.Text,
			Score:  exactExpected + float64(overlap)/float64(denom),
			Answer: answer,
		})
	}
	// Score descending, id ascending — a stable order so a rerun of the
	// same fixture reports the same ranking.
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Score != out[j].Score {
			return out[i].Score > out[j].Score
		}
		return out[i].ID < out[j].ID
	})
	if len(out) > k {
		out = out[:k]
	}
	return out, nil
}

// Run executes every query through the retriever and scores the result.
func Run(f Fixture, retriever Retriever, kValues []int) (Report, error) {
	if len(kValues) == 0 {
		kValues = []int{1, 3, 5, 10}
	}
	maxK := kValues[0]
	for _, k := range kValues {
		if k > maxK {
			maxK = k
		}
	}

	cases := make([]Case, 0, len(f.Queries))
	for _, query := range f.Queries {
		started := time.Now()
		retrieved, err := retriever(query, f, maxK)
		if err != nil {
			return Report{}, err
		}
		latency := float64(time.Since(started).Nanoseconds()) / 1e6

		ids := make([]string, 0, len(retrieved))
		for _, r := range retrieved {
			ids = append(ids, r.ID)
		}
		answer := ""
		for _, r := range retrieved {
			if r.Answer != "" {
				answer = r.Answer
				break
			}
		}
		if answer == "" && len(retrieved) > 0 {
			answer = retrieved[0].Answer
		}
		forbidden := query.ForbiddenMemoryIDs
		if forbidden == nil {
			forbidden = []string{}
		}
		relevant := query.RelevantMemoryIDs
		if relevant == nil {
			relevant = []string{}
		}
		cases = append(cases, Case{
			QueryID:        query.QueryID,
			Category:       query.Category,
			RelevantIDs:    relevant,
			ForbiddenIDs:   forbidden,
			RetrievedIDs:   ids,
			Answer:         answer,
			ExpectedAnswer: query.ExpectedAnswer,
			LatencyMs:      latency,
		})
	}

	retrievalCases := make([]RetrievalCase, 0, len(cases))
	for _, c := range cases {
		retrievalCases = append(retrievalCases, RetrievalCase{
			QueryID: c.QueryID, RelevantIDs: c.RelevantIDs, RetrievedIDs: c.RetrievedIDs,
		})
	}

	var answered []Case
	for _, c := range cases {
		if c.ExpectedAnswer != "" {
			answered = append(answered, c)
		}
	}
	answer := AnswerReport{AnsweredCount: len(answered)}
	if len(answered) > 0 {
		var em, f1 float64
		for _, c := range answered {
			em += ExactMatch(c.Answer, c.ExpectedAnswer)
			f1 += TokenF1(c.Answer, c.ExpectedAnswer)
		}
		answer.ExactMatch = em / float64(len(answered))
		answer.TokenF1 = f1 / float64(len(answered))
	}

	forbiddenHit := map[string]float64{}
	for _, k := range kValues {
		if len(cases) == 0 {
			forbiddenHit[itoa(k)] = 0
			continue
		}
		hits := 0.0
		for _, c := range cases {
			forbidden := map[string]bool{}
			for _, id := range c.ForbiddenIDs {
				forbidden[id] = true
			}
			top := c.RetrievedIDs
			if len(top) > k {
				top = top[:k]
			}
			for _, id := range top {
				if forbidden[id] {
					hits++
					break
				}
			}
		}
		forbiddenHit[itoa(k)] = hits / float64(len(cases))
	}

	latencies := make([]float64, 0, len(cases))
	var sum, max float64
	for _, c := range cases {
		latencies = append(latencies, c.LatencyMs)
		sum += c.LatencyMs
		if c.LatencyMs > max {
			max = c.LatencyMs
		}
	}
	latency := LatencyReport{P95Ms: percentile(latencies, 95), MaxMs: max}
	if len(latencies) > 0 {
		latency.AverageMs = sum / float64(len(latencies))
	}

	return Report{
		Fixture: FixtureSummary{
			Name: f.Name, Kind: f.Kind, Version: f.Version,
			QueryCount: len(f.Queries), MemoryCount: len(f.Memories),
		},
		Retrieval: EvaluateRetrieval(retrievalCases, kValues),
		Answer:    answer,
		Safety:    SafetyReport{ForbiddenHitRateAtK: forbiddenHit},
		Latency:   latency,
		Cases:     cases,
	}, nil
}
